src/array/cuda/rowwise_sampling_prob.cu

```diff
@@ -0,0 +1,657 @@
+  constexpr int TILE_SIZE = 128 / BLOCK_SIZE;
+  if (replace) {  // with replacement.
+    const dim3 block(BLOCK_SIZE);
+    const dim3 grid((num_rows + TILE_SIZE - 1) / TILE_SIZE);
+    CUDA_KERNEL_CALL(
+        (_CSRRowWiseSampleReplaceKernel<IdType, FloatType, TILE_SIZE>),
+        grid, block, 0, stream,
+        rand_seed,
+        num_picks,
+        num_rows,
+        slice_rows,
+        in_ptr,
+        in_cols,
+        data,
+        prob_data,
+        out_ptr,
+        temp_ptr,
+        temp,
+        out_rows,
+        out_cols,
+        out_idxs);
+    device->FreeWorkspace(ctx, temp);
+  } else {  // without replacement
+    IdType* temp_idxs = static_cast<IdType*>(
+        device->AllocWorkspace(ctx, (temp_len) * sizeof(IdType)));
+
+    // Compute A-Res value. A-Res value needs to be calculated only if deg
+    // is greater than num_picks in weighted rowwise sampling without replacement.
+    const dim3 block(BLOCK_SIZE);
+    const dim3 grid((num_rows + TILE_SIZE - 1) / TILE_SIZE);
+    CUDA_KERNEL_CALL(
+        (_CSRAResValueKernel<IdType, FloatType, TILE_SIZE>),
+        grid, block, 0, stream,
+        rand_seed,
+        num_picks,
+        num_rows,
+        slice_rows,
+        in_ptr,
+        data,
+        prob_data,
+        temp_ptr,
+        temp_idxs,
+        temp);
+
+    // sort A-Res value array.
+    FloatType* sort_temp = static_cast<FloatType*>(
+      device->AllocWorkspace(ctx, temp_len * sizeof(FloatType)));
+    IdType* sort_temp_idxs = static_cast<IdType*>(
+      device->AllocWorkspace(ctx, temp_len * sizeof(IdType)));
+
+    cub::DoubleBuffer<FloatType> sort_keys(temp, sort_temp);
+    cub::DoubleBuffer<IdType> sort_values(temp_idxs, sort_temp_idxs);
+
+    void *d_temp_storage = nullptr;
+    size_t temp_storage_bytes = 0;
+    CUDA_CALL(cub::DeviceSegmentedSort::SortPairsDescending(
+      d_temp_storage,
+      temp_storage_bytes,
+      sort_keys,
+      sort_values,
+      temp_len,
+      num_rows,
+      temp_ptr,
+      temp_ptr + 1));
+    d_temp_storage = device->AllocWorkspace(ctx, temp_storage_bytes);
+    CUDA_CALL(cub::DeviceSegmentedSort::SortPairsDescending(
+      d_temp_storage,
+      temp_storage_bytes,
+      sort_keys,
+      sort_values,
+      temp_len,
+      num_rows,
+      temp_ptr,
+      temp_ptr + 1));
+    device->FreeWorkspace(ctx, d_temp_storage);
+    device->FreeWorkspace(ctx, temp);
+    device->FreeWorkspace(ctx, sort_temp);
+
+    // select tok-num_picks as results
+    CUDA_KERNEL_CALL(
+        (_CSRRowWiseSampleKernel<IdType, FloatType, TILE_SIZE>),
+        grid, block, 0, stream,
+        num_picks,
+        num_rows,
+        slice_rows,
+        in_ptr,
+        in_cols,
+        data,
+        out_ptr,
+        temp_ptr,
+        sort_values.Current(),
+        out_rows,
+        out_cols,
+        out_idxs);
+    device->FreeWorkspace(ctx, temp_idxs);
+    device->FreeWorkspace(ctx, sort_temp_idxs);
```

Comment:
device->FreeWorkspace(ctx, temp_idxs);
    device->FreeWorkspace(ctx, sort_temp_idxs);
can move to 610 line